### PR TITLE
feat: Added RedirectAgent.

### PR DESCRIPTION
**Pull request by ShogunPanda**
This PR adds the RedirectAgent as suggested by @mcollina in #553 to add redirections following to undici.

At the moment the PR works only on the top-level `request` method, but I'd like to have your opinion before extending it to `stream` and `pipeline`.

@ronag As requested, the modifications are completely isolated to the agent. I used monkey patching and thus the rest of undici is unaffected.

I have a final question: at the moment, when redirecting, I completely ignore the response body in case 3xx. In theory the server should not send a body but we know how life is :)
My question is: should I call something like `body.end` or `body.close` or can I safely ignore it?

### CODE REVIEW

**Comment by ronag** [comment]
If you override dispatch instead of request then it will automatically work for all of the other methods.

**Comment by ronag** [comment]
Also I think redirect can be quite a bit more complicated https://github.com/follow-redirects/follow-redirects/blob/main/index.js#L308

**Comment by ronag** [comment]
@mcollina: I think the "correct" way to implement this is to extend from `undici.Client` and override `dispatch`, then and somehow pass a factory or prototype to `undici.Pool` which tells it to create that kind of client. Alternatively some kind of hook. wdyt? 

I don't think reimplementing all

**Comment by ronag** [comment]
Actually I think it's sufficient to just create a `RedirectPool` and override `dispatch`:

```js
class RedirectPool extends Pool {
  dispatch (opts, handler) {
    super.dispatch(opts, {
      onHeaders (headers, ...) {
        if (shouldRedirect) {
          this.dispatch(...)
        },
        ...
      }
    })
  }
}
```

**Comment by ronag** [comment]
Even fancier would be just to create a redirect mixin that just override dispatch of any object that implements the client api.

**Comment by ShogunPanda** [comment]
Ok, I'll go for RedirectPool (with mixing) and RedirectAgent.
It will be much more clean!

From `onHeaders` if I return false it will skip the body and trailers part of redirected request, isn't it? Will undici take care of underlying socket or do I have to manage anything?

**Comment by ronag** [comment]
> From `onHeaders` if I return false it will skip the body and trailers part of redirected request, isn't it? Will undici take care of underlying socket or do I have to manage anything?

false will just pause the socket. You have to manage everything unfortunately .

**Comment by mcollina** [comment]
I kind of agree of using a RedirectPool being a better internal implementation rather than monkey patching.

The RedirectPool would need to have a reference to the Agent as Redirects might cross a domains.

We would need to have some logic to avoid never-ending redirect chains (or loops), either in the format of a timeout or as a limit in the number of hops.

**Comment by ShogunPanda** [comment]
Ok, now I remember why I also for Agent: cross origin - Pools or Clients are single origin. 
From my understanding of the code, pool or client don't use agent. The top level method do via `dispatchRequest` internal top level method in `agent.js`. Am I wrong?

@mcollina I already added a `maxRedirects` options with defaults to 10.

@ronag I've checked `follow-redirects` link. Their implementation is weird - On a 301 or 302 you're not obligated to switch to POST to GET. The RFC 7231 says "MAY". So I haven't changed the method in that case.
And I don't even know why they drop the authorization header honestly.
But I do agree that I need to drop `Host` and `Content-*` headers when switching to GET method.

**Comment by mcollina** [comment]
> Ok, now I remember why I also for Agent: cross origin - Pools or Clients are single origin. 
> 
> From my understanding of the code, pool or client don't use agent. The top level method do via `dispatchRequest` internal top level method in `agent.js`. Am I wrong?

You are correct. However monkey patching Pool during get() is not correct as it scrambles a lot of internal implementation of V8. What we are recommending is to change Agent constructor so it accept the Pool class to use. Then, in RedirectAgent, you can set that option to point to a RedirectPool, which implement the redirect logic. Somehow, you'd need to make RedirectPool aware of the agent itselt to support cross origin.

**Comment by ShogunPanda** [comment]
Ok, the first part is trivial.
Let me think a little on how to implement RedirectPool using onHeaders then.

**Comment by mcollina** [comment]
Why did you delete this @ShogunPanda ?

**Comment by ShogunPanda** [comment]
@mcollina I messed when renaming main branch from `master` to `main` in order to be consistent with upstream.

I'm sending the new PR right now.

File: lib/agent.js [hunk 1]
```diff
@@ -41,10 +46,11 @@ class Agent extends EventEmitter {
     let pool = self[kAgentCache].get(origin)
 
     if (!pool) {
-      pool = self[kAgentOpts] && self[kAgentOpts].connections === 1
-        ? new Client(origin, self[kAgentOpts])
-        : new Pool(origin, self[kAgentOpts])
+      const Klass = self[kAgentOpts] && self[kAgentOpts].connections === 1
```
Comment (ronag) [inline lib/agent.js:49]: Klass?

Comment (ShogunPanda) [reply lib/agent.js:49]: Ehm? Can you please clarify?

**Pull request closed without merging**